src/main/java/com/answerdigital/answerking/service/CategoryService.java

```diff
@@ -36,84 +37,173 @@ public CategoryService(final ProductService productService,
-        }
+        final Category category = findById(id);
+        final Category updatedCategory = updateRequestToCategory(category, categoryRequest);
+        addProductsToCategory(updatedCategory, categoryRequest.productIds());
 
-        category.removeProduct(product);
-        final Category savedCategory = categoryRepository.save(category);
-        return categoryMapper.convertCategoryEntityToCategoryResponse(savedCategory);
+        return categoryToResponse(categoryRepository.save(updatedCategory));
     }
 
+    /**
+     * Retires a Category.
+     * @param categoryId The Category ID to retire.
+     */
     public void retireCategory(final Long categoryId) {
         final Category category = findById(categoryId);
         if(category.isRetired()) {
             throw new RetirementException(String.format("The category with ID %d is already retired", categoryId));
         }
         category.setRetired(true);
-        final Category savedCategory = categoryRepository.save(category);
+        categoryRepository.save(category);
     }
 
+    /**
+     * Finds a List of all the Products within a Category.
+     * @param categoryId The ID of the Category.
+     * @return A List of ProductResponses.
+     */
     public List<ProductResponse> findProductsByCategoryId(final Long categoryId) {
         return productService.findProductsByCategoryId(categoryId);
     }
+
+    /**
+     * Adds a list of products to a given category.
+     * @param category The Category entity to add the products to.
+     * @param productIds The List of Product Ids.
+     */
+    @Transactional(propagation = Propagation.REQUIRED)
+    public void addProductsToCategory(final Category category, final List<Long> productIds) {
+        final Set<Product> products = new HashSet<>(productService.findAllProductsInListOfIds(productIds));
+        validateNoProductsAreRetired(products);
+        category.setProducts(products);
+        categoryRepository.save(category);
+    }
+
+    /**
+     * Checks if any products within a Set are retired.
+     * @param products The Set of Products to check.
+     * @throws RetirementException When a product is retired.
+     */
+    private void validateNoProductsAreRetired(final Set<Product> products) {
+        final List<Long> retiredProducts = products
+            .stream()
+            .filter(Product::isRetired)
+            .map(Product::getId)
+            .toList();
+
+        if(!retiredProducts.isEmpty()) {
+            throw new RetirementException(String.format("Products with IDs %s are retired", retiredProducts));
+        }
+    }
+
+    /**
+     * Checks if a Category already exists with a given name.
+     * @param categoryName The Category name.
+     * @param id The ID of the Category.
+     * @throws NameUnavailableException When the Category name already exists.
+     */
+    private void validateCategoryNameDoesNotExist(final String categoryName, final Optional<Long> id) {
```

Comment:
Nested ifs almost always sit wrong with me. I'd argue that we should have this as two separate methods. One for when we're renaming categories and one for when we're creating new ones.

Reply:
Resolved